[pkg/ottl] Support dynamic indexing

odubajDT opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #36644

odubajDT committed:
dollar version

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
different approach in parser

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
working version

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
polish

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
remove dead code

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
revert tests

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
revert tests

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
polish

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
support more use cases

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
fix lint

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
fix lint

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
working tests

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
refactor

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
fix typo

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
make generate

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
support expression as attribute key in parser

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
support converters as attribute keys

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
add chlog

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
polishing of implementation

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
add more e2e tests

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

evan-bradley commented on pkg/ottl/contexts/internal/value.go:
```diff
@@ -71,27 +71,35 @@ func SetValue(value pcommon.Value, val any) error {
 	val := value
 	var ok bool
-	for i := 0; i < len(keys); i++ {
+	for count := 0; count < len(keys); count++ {
```
What does `count` represent here? To me the conventional `i` makes more sense, but I could be missing your intent. `count` strikes me as more of an accumulated/computed value.

odubajDT replied:
`i` was also an integer value retrieved from the getter here, so I renamed the index to avoid confision, but I agree `count` is not the best, will go with `index` here

evan-bradley commented on pkg/ottl/contexts/internal/value.go:
```diff
@@ -164,7 +180,20 @@ func setIndexableValue[K any](ctx context.Context, tCtx K, currentValue pcommon.
 				currentValue = currentValue.Slice().AppendEmpty()
 			default:
-				return errors.New("neither a string nor an int index was given, this is an error in the OTTL")
+				resString, errString := FetchValueFromExpression[K, string](ctx, tCtx, keys[count])
+				resInt, errInt := FetchValueFromExpression[K, int64](ctx, tCtx, keys[count])
```
Would it be possible to take these out of this default case and consider all four cases inside a single switch statement?

odubajDT replied:
Potentially yes, but that would mean we need to fetch values from expression before the switch, which adds complexity with an additional call to simple use-cases like `attributes["index"]` where additional call to `FetchValueFromexpression()` function is not needed. I agree the code looks a little worse, but in end effect it's more efficient

evan-bradley commented on pkg/ottl/functions.go:
```diff
@@ -203,18 +218,42 @@ func (p *basePath[K]) isComplete() error {
+				}
+				par = arg
+			}
+			if f := keys[i].Expression.Float; f != nil {
+				par = literal[K]{value: *f}
+			}
+			if i := keys[i].Expression.Int; i != nil {
+				par = literal[K]{value: *i}
+			}
```
When will these two cases be hit? Would it simplify things if we refactored the grammar to just have keys be a `mathExprLiteral` type instead of a `key` type?

odubajDT replied:
good point, I guess we can leave those out since I cannot hit those with any use-case

Keys need to be also string/int values, otherwise the grammar won't be full

evan-bradley reviewed (commented):
Thanks for taking this one on @odubajDT. I have a few questions on the implementation, but based on the tests passing, I think we're definitely on the right track.

edmocosta commented on pkg/ottl/grammar.go:
```diff
@@ -276,8 +276,9 @@ type field struct {
-	Int    *int64  `parser:"| @Int) ']'"`
+	String     *string          `parser:"'[' (@String "`
+	Int        *int64           `parser:"| @Int"`
+	Expression *mathExprLiteral `parser:"| @@ ) ']'"`
```
That's a nice enhancement! thanks for doing it!

Regarding the grammar changes, we need to ensure that new usages of `mathExprLiteral` (or any other visitable node) are being visited by the `grammarVisitor`, otherwise a few things such as custom errors validation and paths extraction wouldn't work for them. 

For example, without adding the extra logic, the statement `set(attributes[append(attributes["tags"], "prod")], "pass")` would be considered valid, as the second editor (`append`) would be "invisible" to the validator.

I think replacing [grammar.go#L346](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/grammar.go#L346) by a custom `path.accept(v)` method that also goes through fields & keys should do the work :)

edmocosta replied:
That looks good, thanks! I'd just suggest (nitpick) renaming the for _, `arg` vars to something related to the collection it's iterating through.

odubajDT committed:
polishing according to PR review

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
fixing logic

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
polishing variable names

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

evan-bradley commented on pkg/ottl/functions.go:
```diff
@@ -203,18 +218,36 @@ func (p *basePath[K]) isComplete() error {
 	}
 	ks := make([]Key[K], len(keys))
 	for i := range keys {
+		var par Getter[K]
+		if keys[i].Expression != nil {
+			if keys[i].Expression.Path != nil {
+				arg, err := p.buildGetSetterFromPath(keys[i].Expression.Path)
+				if err != nil {
+					return nil, err
+				}
+				par = arg
+			}
+			if keys[i].Expression.Converter != nil {
+				g, err := p.newGetterFromConverter(*keys[i].Expression.Converter)
+				if err != nil {
+					return nil, err
+				}
+				par = g
+			}
+		}
 		ks[i] = &baseKey[K]{
 			s: keys[i].String,
 			i: keys[i].Int,
+			g: par,
```
```suggestion
		var getter Getter[K]
		if keys[i].Expression != nil {
			if keys[i].Expression.Path != nil {
				arg, err := p.buildGetSetterFromPath(keys[i].Expression.Path)
				if err != nil {
					return nil, err
				}
				getter = arg
			}
			if keys[i].Expression.Converter != nil {
				g, err := p.newGetterFromConverter(*keys[i].Expression.Converter)
				if err != nil {
					return nil, err
				}
				getter = g
			}
		}
		ks[i] = &baseKey[K]{
			s: keys[i].String,
			i: keys[i].Int,
			g: getter,
```
Minor, just a variable rename suggestion.

TylerHelmuth reviewed (commented):
This is so exciting! We've wanted this feature for a long time and I'm glad to see that it fit so nicely into our Path/Key pattern.

Can you add e2e tests that use the new ottl-context scoped syntax? I want to ensure statements like 

`set(log.attributes[resource.attributes["flags"]], "something33")` 

work.

odubajDT commented:
> This is so exciting! We've wanted this feature for a long time and I'm glad to see that it fit so nicely into our Path/Key pattern.
> 
> Can you add e2e tests that use the new ottl-context scoped syntax? I want to ensure statements like
> 
> `set(log.attributes[resource.attributes["flags"]], "something33")`
> 
> work.

Should be done, please have a look.

odubajDT committed:
more e2e tests + polishing variable naming

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT commented:
> This is so exciting! We've wanted this feature for a long time and I'm glad to see that it fit so nicely into our Path/Key pattern.
> 
> Can you add e2e tests that use the new ottl-context scoped syntax? I want to ensure statements like
> 
> `set(log.attributes[resource.attributes["flags"]], "something33")`
> 
> work.

Should be done, please have a look

odubajDT committed:
fix lint

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT committed:
fix lint

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

edmocosta reviewed (commented):
I think we should add some extra type validations. Considering the `ottl.Key` index can now be an expression (`mathExprLiteral`), statements like `set(severity_text, attributes[1])` and `set(attributes["slice"][0.0], "bar")` would be considered valid and they would panic on the `FetchValueFromExpression` function. I know they're invalid usages of the OTTL syntax, but I'd prefer to not panic if that happens. WDYT?

odubajDT committed:
cover error cases with invalid index types

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

odubajDT commented:
> I think we should add some extra type validations. Considering the `ottl.Key` index can now be an expression (`mathExprLiteral`), statements like `set(severity_text, attributes[1])` and `set(attributes["slice"][0.0], "bar")` would be considered valid and they would panic on the `FetchValueFromExpression` function. I know they're invalid usages of the OTTL syntax, but I'd prefer to not panic if that happens. WDYT?

You got a point here. Added some additional checks and polished the error messages to be clearer where the problem lies. Also added some e2e tests.

odubajDT committed:
fix lint

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

TylerHelmuth commented on pkg/ottl/e2e/e2e_test.go:
Can you add a test for indexing using a converter and path that return an invalid index value (like a map or a struct or an double or something)

odubajDT replied:
added

odubajDT committed:
more e2e tests

Signed-off-by: odubajDT <ondrej.dubaj@dynatrace.com>

edmocosta reviewed (approved):
Great work! I think in the future we could also change the OTTL contexts/parser to validate the path's key's types, so we can report all invalid usages at startup time instead of run time.
